flowlogs: adding misc fields via configuration

KalmanMeth opened the pull request:
Given the Flow Exporter pipeline developed by @hunchback, we add to the secadvisor pipeline the capability to specify gremlin expressions with substitution in the yml file to generate additional fields in the output flow information. The basic syntax in the yml file looks like:
  transform:
    type: sa
    extend:
      - AA_Name=G.V().Has('RoutingTables.Src','{{.Network.A}}').Values('Host')
      - BB_Name=G.V().Has('RoutingTables.Src','{{.Network.B}}').Values('Host')
This uses the template golang feature, where template expressions are enclosed in {{ }}.
The fields surrounded by `{{ }}` are taken as the names of fields in the flow information and are replaced with their actual values, before evaluating the gremlin expression and placing the result in a new field which is added to the flow information.
@hunchback @eranra Your initial feedback is most valued.
There is also an extension to perform lookup on VM names instead of Container names in file resolve_vm.go.

hunchback commented on contrib/exporters/secadvisor/mod/transform.go:
```diff
@@ -38,7 +38,12 @@ func NewTransform(cfg *viper.Viper) (interface{}, error) {
 	}
 
-	resolver, err = NewResolveMulti(resolver.(Resolver))
+	resolver2, err := NewResolveVm(cfg)
```
I guess you are missing a file - as I don't see the implementation of `NewResolveVM` anywhere ... also if resolution is a specific kind of Vm such as `Xen` then rather call the handler `NewResolveXen`

KalmanMeth replied:
You are right. I forgot to add the file resolve_vm.go.

hunchback commented on contrib/exporters/secadvisor/mod/transform.go:
```diff
@@ -38,7 +38,12 @@ func NewTransform(cfg *viper.Viper) (interface{}, error) {
+		return nil, err
+	}
+
+	resolver, err = NewResolveMulti(resolver.(Resolver), resolver2.(Resolver))
```
now that you have several resolvers then maybe use names such as:
- resolverRunc
- resolverVM (or resolverXen)

etc.

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+func NewExtendNone(cfg *viper.Viper) (interface{}, error) {
+	logging.GetLogger().Infof("entering NewExtendNone")
```
please remove this log message

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+func NewExtendGremlin(cfg *viper.Viper) (interface{}, error) {
+	logging.GetLogger().Infof("entering NewExtendGremlin")
```
remove

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+	logging.GetLogger().Infof("entering NewExtendGremlin")
+	gremlinClient := client.NewGremlinQueryHelper(CfgAuthOpts(cfg))
+	extend := cfg.GetStringSlice(CfgRoot + "transform.extend")
+	logging.GetLogger().Infof("NewExtendGremlin, extend = %s", extend)
```
remove

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/pipeline.go:
```diff
@@ -110,6 +115,11 @@ func init() {
 	FiltererHandlers = make(HandlersMap)
 	FiltererHandlers.Register("subnet", NewFilterSubnet, true)
 
+	logging.GetLogger().Infof("init: before ExtenderHandlers")
```
delete

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/pipeline.go:
```diff
@@ -110,6 +115,11 @@ func init() {
+	logging.GetLogger().Infof("init: before ExtenderHandlers")
+	ExtenderHandlers = make(HandlersMap)
+	ExtenderHandlers.Register("gremlin", NewExtendGremlin, true)
+	logging.GetLogger().Infof("init: after ExtenderHandlers")
```
delete

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/pipeline.go:
```diff
@@ -152,6 +163,13 @@ func NewPipeline(cfg *viper.Viper) (*Pipeline, error) {
 		return nil, err
 	}
 
+	logging.GetLogger().Infof("NewPipeline, before ExtenderHandlers.Init")
```
delete

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/pipeline.go:
```diff
@@ -152,6 +163,13 @@ func NewPipeline(cfg *viper.Viper) (*Pipeline, error) {
+	if err != nil {
+		return nil, err
+	}
+	logging.GetLogger().Infof("NewPipeline, after ExtenderHandlers.Init")
```
delete

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+        return name
+}
+
+// use substitution value and perform gremlin query to obtain the desired data enrichment
```
comment must start with `// PerformConversion `

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+                ii = jj
+        }
+	// TODO: verify that each substring is a valid field so we don't get an exception
+        name := ii[substrings[len(substrings)-1]].(string)
```
put this inside a block:

```
if len(substrings) > 0 {
         
}
```

KalmanMeth replied:
added check earlier that len > 0.

hunchback replied:
ok

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+        return result3[0]
+}
+
+// main function called by pipeline to perform the extension substitutions
```
comment must start with `// Extend ...`

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+// we expect the data of a single flow per call
+func (e *extendGremlin)  Extend(in interface{}) interface{} {
+	// convert the flow information to a map
+	S1, _ := json.Marshal(&in)
```
if we are using other encoding forms (other than JSON) such as CSV then we will get incorrect values (such as wrong field names).

KalmanMeth replied:
Implementation revised to return the original SecurityAdvisorFlow struct.

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+	S1, _ := json.Marshal(&in)
+	var S2 map[string]interface{}
+	S2 = make(map[string]interface{})
+	json.Unmarshal(S1, &S2)
```
doing a json.Marshel() and json.Unmarshel() is very expensive. Typically the Unmarshel() takes longer than Marchel() some benchmarks I noticed a 5x markup :-(

KalmanMeth replied:
We now avoid the marshel/Unmarshel by using the template abstraction.

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+	// parse the extension string and perform the substitution
+        for _, ee := range e.extendStrings {
+		// TODO do this parsing once at the beginning and save in some data structure
+                substrings := strings.Split(ee, "=")
```
you must handling escaping here as the Gremlin expression could contain `=` - for example `g.v().Has("X", "=")`

KalmanMeth replied:
We now split on the first '=', so that any additional '=' is included in the gremlin expression.

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+	// parse the extension string and perform the substitution
+        for _, ee := range e.extendStrings {
+		// TODO do this parsing once at the beginning and save in some data structure
+                substrings := strings.Split(ee, "=")
```
verify that you got exactly 2 strings (key, value)

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+	return S2
+}
+
+
```
redundant newline

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+func (e *extendNone)  Extend(in interface{}) interface{} {
+	return in
```
did you run the code through `gofmt -s -w` ? asking as a notice two-spaces before `Extend`

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+        return &extendNone{}, nil
+}
+
+func (e *extendNone)  Extend(in interface{}) interface{} {
```
did you run the code through `gofmt -s -w` ? asking as a notice two-spaces before `Extend`

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+func (e *extendNone)  Extend(in interface{}) interface{} {
+	return in
```
I don't like the fact that the extendNone returns a `struct` while other extends return a `map` - you really want to align the two cases such that for example conversion to a map is done by the pipeline and Extend() signature is:

```
func Extend(map[string]interface{}) map[string]interface{}
```

KalmanMeth replied:
As noted elsewhere, there is overhead in converting everything to a map in order to perform the substitution, so we want to avoid it if the conversion is unnecessary. Let's discuss how we want to handle this.

hunchback replied:
the alternative suggestion (which helps also solve other issues) is to add an extra field into the SecAdvisorFlow struct which contains a map (this field could be serialized in json both muted for other encoders).

KalmanMeth replied:
We now return the original SecurityAdvisorFlow struct as before.

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+        ii = S
+        for i := 0; i < len(substrings)-1; i++ {
+                str := substrings[i]
+                jj = ii[str].(map[string]interface{})
```
you must add type assertion:

```
val, ok := ii[str]
// handle !ok
jj, ok := val.(map[string]interface{})
// handle !ok
```

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+//   transform:
+//     type: <xxx>
+//     extend:
+//       - VAR_NAME1=<gremlin expression with a single substitution string>
```
calculating a value via a single Gremlin expression will not work in the general case and would even not suffice the SecAdvisor and other cases:

```
// IPToName resolve ip to name
func (r *resolveRunc) IPToName(ipString, nodeTID string) (string, error) {
        node, err := r.gremlinClient.GetNode(g.G.V().Has("Runc.Hosts.IP", ipString).ShortestPathTo(g.Metadata("TID", nodeTID)))
        if err != nil {
                return "", err
        }

        name := node.Metadata["Runc"].(map[string]interface{})["Hosts"].(map[string]interface{})["Hostname"].(string)
        return "0_0_" + name + "_0", nil
}
```



this will not work in the general case - as you may need to run multiple Gremlin for example:

KalmanMeth replied:
Multiple Gremlin expressions can be defined, one after another, to define whatever is needed. Do you mean we need to combine multiple Gremlin expressions in a single query? This is a first implementation with initial functionality. We should discuss what the additional requirements are and see how to address them with improvements over time.

KalmanMeth replied:
To calculate this particular name (not including the prefix 0_0_, etc), you can do:
G.V().Has("Runc.Hosts.IP", ipString).ShortestPathTo(g.Metadata("TID", nodeTID)).Values('Runc.Hosts.Hostname)

hunchback replied:
ok. For now lets' see if we can cover needed functionality w/o requiring multiple Gremlin queries.

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,127 @@
+// use substitution value and perform gremlin query to obtain the desired data enrichment
+func (e *extendGremlin) PerformConversion(origString string, S map[string]interface{}) string {
+	// `%` marks the substitutin string; split the string on `%`
+        substrings := strings.Split(origString, "%")
```
this will not work for the general case of multiple substition values. You must use some type of template library such as: https://github.com/Masterminds/sprig

KalmanMeth replied:
Is this a requirement to have multiple substitutions in a single query? We can expand the functionality over time to cover this case, if needed.

KalmanMeth replied:
Is this a requirement to have multiple substitutions in a single query? We can expand the functionality over time to cover this case, if needed.

hunchback replied:
its' mandatory to be able to express the current resolvers (e.g. runC) using your queries. And as the runC requires both tidNode and IP you must support these. More generally using split according to `%` to replace even a single argument is a very bad idea as it does not account `%` which may be legitimately embedded into the query.

KalmanMeth replied:
We now use templates, so this should be OK.

hunchback commented:
@KalmanMeth - there are some fundamental issues with your approach:

The functionality you offer does not even cover the existing SecAdvisor use case, as for that you work correctly you must also add tidNode value into your query (as IP addresses may not be unique throughout the cluster).

The `encoder` assumes a struct with annotations which then enable it to Marshel to JSON, CSV, <whatever> correctly - by covering to a map (using JSON serialize - you kill CSV and other non-JSON encoders).

You performance is poor - I assume an easy 10x slowdown due to using json.Marshel/json.Unmarshel for convert to a map - an alternative option would  to add a field within SecAdvisorFlow which contains a `map[string]interface{}` and then add fields to there.

KalmanMeth commented:
We now use the template abstraction, as suggested. This avoids the Marshel/Unmarshel. Caching was added. We should add the ability for the user to guide us whether we may use caching.

hunchback commented:
@KalmanMeth - please squash all commits together its' also fine if you split your work  to several commits as long as each commit represents a logical step and does not introduce degredations (thus is stable)

hunchback commented:
@KalmanMeth - regarding testing you can generate a configuration object within your code as demonstrated here https://github.com/skydive-project/skydive/blob/fe4bd6815e1533e0027f0f9b32829f98624f3e91/tests/tests.go#L59

KalmanMeth commented:
I modified transform_test.go to include support for the extensions with misc fields and added a unit test to see that the substitution is done properly.

eranra commented:
@dubek  --> I want to make sure that the new feature that @KalmanMeth is adding in this PR supports the new format that David told me you are implementing for SA. Can you (a) review this PR and (b) work with Kalman on configuration for this PR that fits the new extended fields in format for SA. THX

dubek commented:
I'm on vacation, I'll take a look next week.

On Sun, Aug 25, 2019, 16:05 Eran Raichstein <notifications@github.com>
wrote:

> @dubek <https://github.com/dubek> --> I want to make sure that the new
> feature that @KalmanMeth <https://github.com/KalmanMeth> is adding in
> this PR supports the new format that David told me you are implementing for
> SA. Can you (a) review this PR and (b) work with Kalman on configuration
> for this PR that fits the new extended fields in format for SA. THX
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/skydive-project/skydive/pull/1955?email_source=notifications&email_token=AAACH6YRCP6LVUSNAMMLJ2DQGJ7TLA5CNFSM4ILUP5XKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD5CTHSI#issuecomment-524628937>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAACH62K4GHEGLAWMFSHER3QGJ7TLANCNFSM4ILUP5XA>
> .
>

hunchback commented on contrib/exporters/secadvisor/mod/transform.go:
```diff
@@ -38,7 +38,12 @@ func NewTransform(cfg *viper.Viper) (interface{}, error) {
 	}
 
-	resolver, err = NewResolveMulti(resolver.(Resolver))
+	resolver2, err := NewResolveVm(cfg)
```
you have not added NewResolveVm() to this commit ... so this will not build

KalmanMeth replied:
done

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,160 @@
+/*
```
the commit comment is a mess ... please fix

KalmanMeth replied:
This file (core/extend.go) should have been removed. Please ignore code in this file.

hunchback commented on contrib/exporters/core/store_s3.go:
```diff
@@ -202,6 +203,7 @@ func (s *StoreS3) flushFlowsToObject(t Tag, endTime time.Time) error {
 		return err
 	}
 
+	// XXX use .gz suffix only if the compression was gzip
```
rather use FIXME (not XXX)

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,160 @@
+	"fmt"
+	"time"
+	"encoding/json"
+	"github.com/skydive-project/skydive/api/client"
```
you want to place an empty line between std imports and others

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,160 @@
+import (
+	"fmt"
+	"time"
+	"encoding/json"
```
order is wrong ..
 please run through gofmt

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,160 @@
+type extendNone struct {
+}
+
+func NewExtendNone(cfg *viper.Viper) (interface{}, error) {
```
add comment - // NewExtendNone ....

KalmanMeth replied:
This file (core/extend.go) should have been removed. Please ignore code in this file.

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,160 @@
+
+// extendGremlin is used for transform extensions using gremlin expressions
+// Syntax supported is:
+//	VAR_NAME=<gremlin expression with a single substitution string>
```
wrong

KalmanMeth replied:
This is an old version of the code. Not sure why the new code doesn't show up here.

KalmanMeth replied:
This file (core/extend.go) should have been removed. Please ignore code in this file.

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,160 @@
+// Syntax supported is:
+//	VAR_NAME=<gremlin expression with a single substitution string>
+// For example:
+//	AA_Name=G.V().Has('RoutingTables.Src',%Network.A%).Values('Host')
```
wrong

KalmanMeth replied:
This file (core/extend.go) should have been removed. Please ignore code in this file.

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+// DeriveValue: use the substitution string to extract the value from the provided map
+func (e *extendGremlin) DeriveValue(S map[string]interface{}, varString string) (string, error) {
```
should be a private method deriveValue()

KalmanMeth replied:
This file (core/extend.go) should have been removed. Please ignore code in this file.

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,160 @@
+		str := substrings[i]
+		jj, ok = ii[str].(map[string]interface{})
+		if !ok {
+			logging.GetLogger().Errorf("substring %s extended gremlin not found", str)
```
remove log if u are returning error to caller

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,160 @@
+	}
+	name, ok := ii[substrings[len(substrings)-1]].(string)
+	if !ok {
+		logging.GetLogger().Errorf("substring %s extended gremlin not found", substrings[len(substrings)-1])
```
remove log if u are returning error to caller

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/secadvisor/mod/transform_test.go:
```diff
@@ -130,9 +203,10 @@ func Test_Transform_basic_flow(t *testing.T) {
 }
 
-func Test_Transform_UpdateCount_increses(t *testing.T) {
+func Test_Transform_UpdateCount_increases(t *testing.T) {
```
in go the standard notation is CamelCase so name should rather be TestTransformUpdateCountIncreases()

hunchback commented on contrib/exporters/secadvisor/mod/transform_test.go:
```diff
@@ -150,3 +225,15 @@ func Test_Transform_Status_updates(t *testing.T) {
 	assertEqual(t, "ENDED", secAdvFlow.Status)
 }
+
+func Test_Extend_Gremlin(t *testing.T) {
```
same

hunchback commented on contrib/exporters/secadvisor/mod/transform_test.go:
```diff
@@ -22,14 +22,70 @@ import (
+	return err
+}
+
+func (e *extendGremlin) populateExtendGremlin(f *flow.Flow, transformer *securityAdvisorFlowTransformer) {
```
make a helper func testResolveToHost() to be called for A name and B name

KalmanMeth replied:
Not sure what you mean here.

hunchback replied:
notice that you are performing the same 3 operations per Network.A and Network.B - rather use a helper function

hunchback commented on contrib/exporters/secadvisor/mod/transform_test.go:
```diff
@@ -22,14 +22,70 @@ import (
 	"github.com/skydive-project/skydive/flow"
 )
 
+const testConfig = `---
```
you can remove all but the pipeline.transform section

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/secadvisor/mod/transform_test.go:
```diff
@@ -22,14 +22,70 @@ import (
+	"github.com/skydive-project/skydive/config"
 	"github.com/skydive-project/skydive/common"
-	"github.com/skydive-project/skydive/contrib/exporters/core"
+	//"github.com/skydive-project/skydive/contrib/exporters/core"
```
delete

KalmanMeth replied:
OK

hunchback commented on contrib/exporters/core/extend.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+// PerformConversion: use substitution value and perform gremlin query to obtain the desired data enrichment
+func (e *extendGremlin) PerformConversion(origString string, S map[string]interface{}) (string, error) {
```
remove func

KalmanMeth replied:
This file (core/extend.go) should have been removed. Please ignore code in this file.

KalmanMeth commented on contrib/exporters/secadvisor/mod/resolve_vm.go:
```diff
@@ -0,0 +1,64 @@
+	return node.Host, nil
+}
+
+// IPToName resolve tid to type
```
OK

hunchback commented:
@KalmanMeth - please organize you commits - I expect to still have only 3 commits - also please fix the commit comments

hunchback commented on contrib/exporters/secadvisor/mod/resolve_runc.go:
```diff
@@ -44,6 +45,9 @@ func (r *resolveRunc) IPToName(ipString, nodeTID string) (string, error) {
 	if err != nil {
 		return "", err
 	}
+	if (node == nil) {
```
redundant as gremlinClinet.GetNode() already addresses this case

KalmanMeth replied:
OK

hunchback commented:
please rebase

hunchback commented on contrib/exporters/secadvisor/mod/extend_gremlin.go:
```diff
@@ -0,0 +1,125 @@
+	"github.com/spf13/viper"
+)
+
+// extendGremlin is used for transform extensions using gremlin expressions
```
I don't think you need such a big comment - what is obvious from the code does not require explanations

hunchback commented:
@KalmanMeth - please split into two PRs:
- adding VM resolver
- adding misc extension of fields
Each will be addressed separately

dubek commented:
One more idea:

For the key names (variable names) you currently use strings which will be keys under `"Extend"`. Instead, maybe we can give a JSONPath like:  `.Extend.my_var_name`. This will also allow to add stuff in other places in the JSON, for example in: `.Context.A.Image` or wherever.

hunchback commented:
please remove the VM resolver from here

KalmanMeth commented:
The VM resolver appears because this branch is checked-out from the resolve_vm branch, which has this "adding VM resolver" commit.

KalmanMeth commented:
If I remove it, I have to rebase again after the resolve_vm PR is merged.

hunchback commented:
> If I remove it, I have to rebase again after the resolve_vm PR is merged.

no. these efforts are disjoint, remove the resolve-vm commit and rebase on origin/master

KalmanMeth commented:
I removed the relove_vm from this PR.

KalmanMeth commented:
PR moved to skydive-flow-exporters repository, PR 18. 
https://github.com/skydive-project/skydive-flow-exporter/pull/18
